### DISCUSSION TITLE
What's the best way to validate model attributes on a per attribute basis?

### DISCUSSION THREAD

**Discussion by SeanCoaker** [pydantic/pydantic#8637] [Question]
Hi,

**TL;DR:** What's the best way to validate model attributes on a per attribute basis after having already used model_construct to instantiate the model?

I'm doing some work utilising Pydantic V2. I have some models that I pass data to with the sole intention of validating that data before putting it all into a long string to be passed to the front end. I don't want to go down the usual route with Pydantic of initialising a model and waiting for a ValidationError to be raised. Instead, I'm using model_construct to create an instance of the model, before then trying to validate the model after. The motivation behind this is so that I can detect which parts of the data are malformed, and mark them as being malformed when passing the text back to the front end. In other words, I still want the model instance to store the data being passed in, whether there is a ValidationError or not.

On top of this, I don't want to run model_validate later due to atomicity issues. For example, if I run model_validate, it correctly coerces string inputs to ints, but when a ValidationError is detected, that coercion work is undone.

That leads me to my current solution, which is using BaseModel.__pydantic_validator__.validate_assignment. So I currently run over each attribute and value within my model, and run validate_assignment whilst keeping track of which attributes failed validation. 

My question is, is there a nicer/more efficient way of doing this? This solution works for me but I was wondering if there was a more API-friendly/more robust solution.

Thanks.
